File: core/src/main/java/ru/pocketbyte/locolaser/utils/JsonParseUtils.java
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.utils;

import ru.pocketbyte.locolaser.exception.InvalidConfigException;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

import java.io.File;
import java.io.IOException;
import java.util.List;

/**
 * Utils class to parse JSON.
 *
 * @author Denis Shurygin
 */
public class JsonParseUtils {

    public static final JSONParser JSON_PARSER = new JSONParser();

    public static String getString(JSONObject json, String key, String parentKey, boolean throwIfNull) throws InvalidConfigException {
        Object object = getObject(json, key, parentKey, throwIfNull);
        if (object != null && !(object instanceof String))
            throw new InvalidConfigException("Property " + keyName(key, parentKey) + " must be a String.");
        return (String) object;
    }

    public static boolean getBoolean(JSONObject json, String key, String parentKey, boolean throwIfNull) throws InvalidConfigException {
        Object object = getObject(json, key, parentKey, throwIfNull);
        if (object != null && !(object instanceof Boolean))
            throw new InvalidConfigException("Property " + keyName(key, parentKey) + " must be a Boolean.");
        return object != null ? (Boolean) object : false;
    }

    public static long getLong(JSONObject json, String key, String parentKey, boolean throwIfNull) throws InvalidConfigException {
        Object object = getObject(json, key, parentKey, throwIfNull);
        if (object != null && !(object instanceof Long))
            throw new InvalidConfigException("Property " + keyName(key, parentKey) + " must be a Long.");
        return object != null ? (Long) object : 0;
    }

    public static JSONObject getJSONObject(JSONObject json, String key, String parentKey, boolean throwIfNull) throws InvalidConfigException {
        Object object = getObject(json, key, parentKey, throwIfNull);
        if (object != null && !(object instanceof JSONObject))
            throw new InvalidConfigException("Property " + keyName(key, parentKey) + " must be a JSONObject.");
        if (object != null)
            return (JSONObject) object;
        return null;
    }

    public static List<String> getStrings(JSONObject json, String key, String parentKey, boolean throwIfNull) throws InvalidConfigException {
        Object object = getObject(json, key, parentKey, throwIfNull);
        if (object != null) {
            if(object instanceof JSONArray) {
                for (Object item: (JSONArray) object)
                    if (!(item instanceof String))
                        throw new InvalidConfigException("Property " + keyName(key, parentKey) + " must be a Strings array.");
                return (List<String>) object;
            }
            else
                throw new InvalidConfigException("Property " + keyName(key, parentKey) + " must be a Strings array.");
        }
        return null;
    }

    public static File getFile(JSONObject json, String key, String parentKey, boolean throwIfNull) throws InvalidConfigException {
        String path = getString(json, key, parentKey, throwIfNull);
        if (path != null)
            try {
                return new File(new File(path).getCanonicalPath());
            } catch (IOException e) {
                e.printStackTrace();
            }
        return null;
    }

    public static Object getObject(JSONObject json, String key, String parentKey, boolean throwIfNull) throws InvalidConfigException {
        Object object = json.get(key);
        if (throwIfNull && object == null)
            throw new InvalidConfigException("Property " + keyName(key, parentKey) + " is not set.");
        return object;
    }

    public static String keyName(String key, String parent) {
        StringBuilder builder = new StringBuilder("\"");
        if (parent != null)
            builder.append(parent).append(".");
        return builder.append(key).append("\"").toString();
    }
}

File: core/src/main/java/ru/pocketbyte/locolaser/utils/JsonParseUtils.kt
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.utils

import ru.pocketbyte.locolaser.exception.InvalidConfigException
import org.json.simple.JSONArray
import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser

import java.io.File
import java.io.IOException

/**
 * Utils class to parse JSON.
 *
 * @author Denis Shurygin
 */
object JsonParseUtils {

    val JSON_PARSER = JSONParser()

    @Throws(InvalidConfigException::class)
    fun getString(json: JSONObject, key: String, parentKey: String?, throwIfNull: Boolean): String? {
        val jsonObject = getObject(json, key, parentKey, throwIfNull)
        if (jsonObject != null && jsonObject !is String)
            throw InvalidConfigException("Property ${keyName(key, parentKey)} must be a String.")
        return jsonObject as String?
    }

    @Throws(InvalidConfigException::class)
    fun getBoolean(json: JSONObject, key: String, parentKey: String?, throwIfNull: Boolean): Boolean {
        val jsonObject = getObject(json, key, parentKey, throwIfNull)
        if (jsonObject != null && jsonObject !is Boolean)
            throw InvalidConfigException("Property ${keyName(key, parentKey)} must be a Boolean.")
        return (jsonObject as? Boolean) ?: false
    }

    @Throws(InvalidConfigException::class)
    fun getLong(json: JSONObject, key: String, parentKey: String?, throwIfNull: Boolean): Long {
        val jsonObject = getObject(json, key, parentKey, throwIfNull)
        if (jsonObject != null && jsonObject !is Long)
            throw InvalidConfigException("Property ${keyName(key, parentKey)} must be a Long.")
        return (jsonObject as? Long) ?: 0L
    }

    @Throws(InvalidConfigException::class)
    fun getJSONObject(json: JSONObject, key: String, parentKey: String?, throwIfNull: Boolean): JSONObject? {
        val jsonObject = getObject(json, key, parentKey, throwIfNull)
        if (jsonObject != null && jsonObject !is JSONObject)
            throw InvalidConfigException("Property ${keyName(key, parentKey)} must be a JSONObject.")
        return jsonObject as? JSONObject
    }

    @Throws(InvalidConfigException::class)
    fun getStrings(json: JSONObject, key: String, parentKey: String?, throwIfNull: Boolean): List<String>? {
        val jsonObject = getObject(json, key, parentKey, throwIfNull)
        if (jsonObject != null) {
            if (jsonObject is JSONArray) {
                (jsonObject as JSONArray?)!!
                        .filter { it !is String }
                        .forEach { _ ->
                            throw InvalidConfigException("Property ${keyName(key, parentKey)} must be a Strings array.")
                        }
                return jsonObject as? List<String>
            } else
                throw InvalidConfigException("Property ${keyName(key, parentKey)} must be a Strings array.")
        }
        return null
    }

    @Throws(InvalidConfigException::class)
    fun getFile(json: JSONObject, key: String, parentKey: String?, throwIfNull: Boolean): File? {
        val path = getString(json, key, parentKey, throwIfNull)
        if (path != null)
            try {
                return File(File(path).canonicalPath)
            } catch (e: IOException) {
                e.printStackTrace()
            }

        return null
    }

    @Throws(InvalidConfigException::class)
    fun getObject(json: JSONObject, key: String, parentKey: String?, throwIfNull: Boolean): Any? {
        val jsonObject = json[key]
        if (throwIfNull && jsonObject == null)
            throw InvalidConfigException("Property ${keyName(key, parentKey)} is not set.")
        return jsonObject
    }

    private fun keyName(key: String, parent: String?): String {
        val builder = StringBuilder("\"")
        if (parent != null)
            builder.append(parent).append(".")
        return builder.append(key).append("\"").toString()
    }
}
